Ml flow

mizaqq opened the pull request.

adamwawrzynski commented on model/main.py:
```diff
@@ -1,21 +1,20 @@
 def parse_args() -> argparse.Namespace:
     argparser = argparse.ArgumentParser()
-    argparser.add_argument("--model", type=str, default=ModelType.REGRESSOR)
+    argparser.add_argument("--model", type=str, default=ModelType.XGB)
```
Is this necessary in this PR? I think it might be an artifact for experimenting with different models.

adamwawrzynski commented on model/main.py:
```diff
@@ -29,24 +28,30 @@ def parse_args() -> argparse.Namespace:
 
 def main(args):
-    df = preprocess_data(pd.read_csv(args.train_data))
+    mlflowhandler = MLFlowHandler()
```
Use snake case.
```suggestion
    mlflow_handler = MLFlowHandler()
```

adamwawrzynski commented on model/main.py:
```diff
@@ -29,24 +28,30 @@ def parse_args() -> argparse.Namespace:
-    print(model.evaluate(X_test, y_test))
+    score = model.evaluate(X_test, y_test)
+    mlflowhandler.log_model(model, ("r2_score", score), params, args.train_data, X_train)
+    mlflowhandler.close()
```
Instead of manually logging parameters I would re-write this functionality as a method in the `Model` class with inheriting classes having a common interface to get model parameters.

adamwawrzynski commented on model/utils/data_analysis.py:
```diff
@@ -40,8 +41,13 @@ def get_missing_values(df: pd.DataFrame) -> List[tuple]:
     missing_values_col = get_missing_values(df)
+    mlflow.log_analysis(uncorrelated_col, "correlation coefficient")
+    mlflow.log_analysis(insignificant_col, f"chi square p-value Threshhold: {treshhold}")
+    mlflow.log_analysis(missing_values_col, "missing values")
```
This function is hard to test right now. You should apply S from SOLID principles - "Single responsibility". This function has 2 responsibilities: it computes different kinds of features AND uploads them to the MLFlow. Separate those 2 functionalities into 2 functions.

adamwawrzynski commented on model/utils/data_preprocess.py:
```diff
@@ -62,8 +61,10 @@ def encode_onehot(df: pd.DataFrame) -> pd.DataFrame:
 
-def preprocess_data(df: pd.DataFrame, df_test: Optional[pd.DataFrame] = None) -> pd.DataFrame:
-    uncorrelated_col, insignificant_col, missing_values_col = get_data_for_preprocessing(df, treshhold=0.05)
+def preprocess_data(mlflow: MLFlowHandler, df: pd.DataFrame, df_test: Optional[pd.DataFrame] = None) -> pd.DataFrame:
+    uncorrelated_col, insignificant_col, missing_values_col = get_data_for_preprocessing(
+        df, treshhold=0.05, mlflow=mlflow
+    )
```
Same here. Right now, this function is only used in the notebook. I would remove anything related to the MLFlow in this function, add an additional line to the notebook right after the `preprocess_data` function call, and log those values.

mizaqq replied:
This function is also used in main function that logs preprocessed data to MLFlow for every iteration.
Preprocess data calls the get_data_for_preprocessing(this logs the values used for preprocessing). After preprocessing the preprocessed data is logged to mlflow. 
Then I would need more information on how to rewrite this. This data shouldn't be logged?

adamwawrzynski replied:
Can't you run it once before training and then train the model on the preprocessed data? It's good practice as you can then cache those preprocessed values and speed up the training.

What I was suggesting before is to remove the first argument `mlflow: MLFlowHandler` from the function call and then just directly use the API of this object to upload the results of this function. This is the separation of concerns. Right now this function does 2 things - preprocessing the data AND uploading them to the MLFlow. This might have been 2 functions.

adamwawrzynski commented on model/utils/mlflow.py:
```diff
@@ -0,0 +1,69 @@
+        mlflow.login()
+        mlflow.set_tracking_uri(uri)
+        mlflow.set_experiment(experiment_name)
+        mlflow.set_registry_uri("databricks-uc")
```
Change hardcoded values to the argument with a default value.

adamwawrzynski commented on model/utils/mlflow.py:
```diff
@@ -0,0 +1,69 @@
+    def log_preprocessed_data(self, df: pd.DataFrame) -> None:
+        df.to_csv("preprocessed_data.csv")
+        mlflow.log_artifact("preprocessed_data.csv")
+        os.remove("preprocessed_data.csv")
```
Same here. Change this to the argument with the default value.

adamwawrzynski commented on model/utils/mlflow.py:
```diff
@@ -0,0 +1,69 @@
+
+    def log_analysis(self, data: List[Tuple], metric_type: str) -> None:
+        for item in data:
+            mlflow.log_metric(item[0] + " " + metric_type, item[1])
```
If this tuple has only 2 elements I would change the typing of the `data` parameter to the `List[Tuple[int, int]]`.

adamwawrzynski commented on model/utils/mlflow.py:
```diff
@@ -0,0 +1,69 @@
+        score: Tuple[str, float],
+        params: dict = None,
+        artifact: Path = None,
+        X_train: pd.DataFrame = None,
```
```suggestion
        params: Optional[dict] = None,
        artifact: Optional[Path] = None,
        X_train: Optional[pd.DataFrame] = None,
```

adamwawrzynski reviewed (commented):
Couple of suggestions:
- You could change the PR title to something more informative at first sight: "Add MLFlow support"
- Empty description
- Commit messages are not meaningful. Try to use the git messaging scheme: https://gist.github.com/abravalheri/34aeb7b18d61392251a2

mizaqq commented:
The main changes are
- Docker compose for setting up MLFlow in cloud env
- Refactored model.get_params() as an interface
- Added type hints to the code
- Refactored get_data_for_preprocessing to follow SOLID principles

adamwawrzynski commented on mlflow/docker-compose.yaml:
```diff
@@ -0,0 +1,15 @@
+      - 5000:5000
+    environment:
+      MLFLOW_SERVER_DEFAULT_ARTIFACT_ROOT: /mlflow/artifacts
+    command: mlflow server --backend-store-uri postgresql://postgres:password@db:5432/mlflow --host 0.0.0.0
+  db:
+    image: postgres:latest
+    environment:
+      POSTGRES_USER: postgres
+      POSTGRES_PASSWORD: password
+      POSTGRES_DB: mlflow
```
You should never upload credentials to the repository. I know this is just a pet project let's make a good habit regarding secrets. Common practice while working with `docker-compose` is to use environment variables with all confidential values. You can pass those values from CLI or provide `.env` file. Documentation: https://docs.docker.com/compose/how-tos/environment-variables/set-environment-variables/.

adamwawrzynski commented on model/models/models.py:
```diff
@@ -35,6 +35,10 @@ def predict(self, X: pd.DataFrame) -> pd.DataFrame:
         pass
 
+    @abstractmethod
+    def get_params(self):
```
Missing return type hint.

adamwawrzynski commented on model/models/models.py:
```diff
@@ -49,6 +53,9 @@ def predict(self, X: pd.DataFrame) -> pd.DataFrame:
     def evaluate(self, X, y: pd.DataFrame) -> float:
         return r2_score(y, self.predict(X))
 
+    def get_params(self):
```
And here.

adamwawrzynski commented on model/models/models.py:
```diff
@@ -64,6 +71,9 @@ def predict(self, X: pd.DataFrame) -> pd.DataFrame:
     def evaluate(self, X, y: pd.DataFrame) -> float:
         return r2_score(y, self.predict(X))
 
+    def get_params(self):
```
And here.

adamwawrzynski commented on model/models/models.py:
```diff
@@ -83,3 +93,6 @@ def predict(self, X: pd.DataFrame) -> pd.DataFrame:
     def evaluate(self, X, y: pd.DataFrame) -> float:
         return r2_score(y, self.predict(X))
+
+    def get_params(self):
```
And here.

adamwawrzynski commented on model/utils/mlflow.py:
```diff
@@ -17,29 +15,27 @@ def close(self):
 
     def login_to_mlflow(
-        self, uri: str = "databricks", experiment_name: str = "/Users/michal.zareba@softwaremill.pl/housepricesml"
+        self, uri: str = "http://48.209.80.111:5000", experiment_name: str = "housepricesml"
```
I would suggest having this in constructor of an object. And second thing, I would recommend fetching this data from Environment variables or from config file, e.g. `uri = os.environ.get("MLFLOW_TRACKING_URI", "http://48.209.80.111:5000")`

adamwawrzynski commented on model/utils/mlflow.py:
```diff
@@ -17,29 +15,27 @@ def close(self):
-        mlflow.log_artifact("preprocessed_data.csv")
-        os.remove("preprocessed_data.csv")
+    def log_preprocessed_data(self, df: pd.DataFrame, path: Path = Path("preprocessed_data.csv")) -> None:
+        df.to_csv(path)
+        mlflow.log_artifact(path)
+        os.remove(path)
```
Here you are providing a path to the file that you are creating and then deleting. This is the perfect case for using temporary files. 

```python
import tempfile

with tempfile.NamedTemporaryFile() as file:
    ...
```

adamwawrzynski reviewed (changes requested):
Remove credentials from the repository. Overall, LGTM.

adamwawrzynski commented on mlflow/.env:
```diff
@@ -0,0 +1,3 @@
+POSTGRES_USER=postgres
+POSTGRES_PASSWORD=password
+POSTGRES_DB=mlflow
```
This file should not be added to the repository. Eventually, you can add this file with empty values to guide users of this repository on which variables should be provided.

adamwawrzynski commented on mlflow/docker-compose.yaml:
```diff
@@ -0,0 +1,12 @@
+      - 5000:5000
+    environment:
+      MLFLOW_SERVER_DEFAULT_ARTIFACT_ROOT: /mlflow/artifacts
+    command: mlflow server --backend-store-uri postgresql://postgres:password@db:5432/mlflow --host 0.0.0.0
```
Add `POSTGRES_PORT` variable to the `.env` file.
```suggestion
    command: mlflow server --backend-store-uri postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@db:${POSTGRES_PORT}/mlflow --host 0.0.0.0
```

adamwawrzynski commented on mlflow/docker-compose.yaml:
```diff
@@ -0,0 +1,12 @@
+    build: .
+    image: ghcr.io/mlflow/mlflow
+    ports:
+      - 5000:5000
```
You should add variable `MLFLOW_PORT` to the `.env` file to do that.
```suggestion
      - ${MLFLOW_PORT}:5000
```

adamwawrzynski commented on model/main.py:
```diff
@@ -27,31 +28,35 @@ def parse_args() -> argparse.Namespace:
-    mlflowhandler.log_model(model, ("r2_score", score), params, args.train_data, X_train)
-    mlflowhandler.close()
+    mlflow_handler.log_model(model, ("r2_score", score), X_train, model.get_params(), args.train_data)
+    mlflow_handler.close()
```
This is just a suggestion. You can implement a context manager and use `with` statement. Then you have certainty that resources are freed after usage and don't have to remember about having `.close()` method call after.

adamwawrzynski reviewed (changes requested):
Fix `.env` file with credentials in the repo and accessing ENV variables in the `docker-compose` files. Context manager issue is just a suggestion.

adamwawrzynski commented on mlflow/.env:
```diff
@@ -1,3 +1,5 @@
 POSTGRES_PASSWORD=password
 POSTGRES_DB=mlflow
+POSTGRES_PORT=5432
+MLFLOW_PORT=5000
```
Remove this file from repository.

mizaqq replied:
deleted [c6e7654](https://github.com/mizaqq/housepricesml/pull/3/commits/c6e765427b57568b8622b9d4233846e7aa5050ea)
